code/_onclick/item_attack.dm

```diff
@@ -1,155 +1,242 @@
+	// Attack phase
 
-// Called when the item is in the active hand, and clicked; alternately, there is an 'activate held object' verb or you can hit pagedown.
-/obj/item/proc/attack_self(mob/user)
-	var/signal_ret = SEND_SIGNAL(src, COMSIG_ITEM_ATTACK_SELF, user)
-	if(signal_ret & COMPONENT_NO_INTERACT)
+	if(pre_attack(target, user, params))
 		return
-	if(signal_ret & COMPONENT_CANCEL_ATTACK_CHAIN)
-		return TRUE
 
-/obj/item/proc/pre_attack(atom/A, mob/living/user, params) //do stuff before attackby!
-	if(SEND_SIGNAL(src, COMSIG_ITEM_PRE_ATTACK, A, user, params) & COMPONENT_CANCEL_ATTACK_CHAIN)
-		return TRUE
+	// Deal with objects which haven't been migrated to the new attack chain.
+	// This is the only stanza here that's allowed to call legacy code.
+	if(target.new_attack_chain)
+		if(!target.attack_by(src, user, params))
+			finish_attack_chain(user, target, proximity_flag, params)
+	else
+		var/resolved = target.attackby__legacy__attackchain(src, user, params)
+		if(!resolved && target && !QDELETED(src))
+			if(new_attack_chain)
+				finish_attack_chain(user, target, proximity_flag, params)
+			else
+				afterattack__legacy__attackchain(target, user, proximity_flag, params) // 1: clicking something Adjacent
+
+/// Post-attack behavior for objects using the new attack chain.
+/obj/item/proc/finish_attack_chain(mob/user, atom/target, proximity_flag, params)
+	PRIVATE_PROC(TRUE)
 
-	if(SEND_SIGNAL(A, COMSIG_ITEM_BEING_ATTACKED, src, user, params) & COMPONENT_CANCEL_ATTACK_CHAIN)
+	// At this point it means the attack was "successful", or at least
+	// unhandled, in some way. This can mean nothing happened, this can mean the
+	// target took damage, etc.
+	target.attacked_by(src, user, params)
+
+	// Finally deal with anything that should happen after a successful attack.
+	SEND_SIGNAL(src, COMSIG_AFTER_ATTACK, target, user, proximity_flag, params)
+	SEND_SIGNAL(target, COMSIG_AFTER_ATTACKED_BY, src, user, proximity_flag, params)
+
+	after_attack(target, user, proximity_flag, params)
+
+/// Called when the item is in the active hand, and clicked; alternately, there
+/// is an 'activate held object' verb or you can hit pagedown.
+/obj/item/proc/activate_self(mob/user)
+	SHOULD_CALL_PARENT(TRUE)
+
+	if(SEND_SIGNAL(src, COMSIG_ACTIVATE_SELF, user) & COMPONENT_CANCEL_ATTACK_CHAIN)
+		return FINISH_ATTACK
+
+/**
+ * Called on ourselves before we hit something. Return TRUE to cancel the remainder of the attack chain.
+ *
+ * Arguments:
+ * * atom/A - The atom about to be hit
+ * * mob/living/user - The mob doing the htting
+ * * params - click params such as alt/shift etc
+ *
+ * See: [/obj/item/proc/melee_attack_chain]
+ */
+/obj/item/proc/pre_attack(atom/A, mob/living/user, params)
+	SHOULD_CALL_PARENT(TRUE)
+
+	if(SEND_SIGNAL(src, COMSIG_PRE_ATTACK, A, user, params) & COMPONENT_CANCEL_ATTACK_CHAIN)
 		return TRUE
 
+	// TODO: Turn this into a component and have a sane implementation instead of extra-specific behavior in a core proc
```

Comment:
Here's a TODO, plan on changing it in this PR or during the next steps?

Reply:
Next steps. I don't want to risk having to test more things in the initial change.